src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -3,45 +3,418 @@
+
+                                    // Statics for valuetypes where the valuetype is defined in this module are handled here. 
+                                    // Other valuetype statics utilize the pessimistic model below.
+                                case CorElementType.ELEMENT_TYPE_VALUETYPE:
+                                    isGcField = true;
+                                    alignment = pointerSize;
+                                    size = pointerSize;
+                                    break;
+
+                                case CorElementType.ELEMENT_TYPE_END:
+                                default:
+                                    isGcField = true;
+                                    alignment = pointerSize;
+                                    size = pointerSize;
+                                    if (!valueTypeHandle.IsNil)
+                                    {
+                                        // Allocate pessimistic non-GC area for cross-module fields as that's what CoreCLR does
+                                        // <a href="https://github.com/dotnet/coreclr/blob/659af58047a949ed50d11101708538d2e87f2568/src/vm/ceeload.cpp#L2124">here</a>
+                                        nonGcStatics[index] = LayoutInt.AlignUp(nonGcStatics[index], new LayoutInt(TargetDetails.MaximumPrimitiveSize))
+                                            + new LayoutInt(TargetDetails.MaximumPrimitiveSize);
+                                    }
+                                    else
+                                    {
+                                        // Field has an unexpected type
+                                        throw new InvalidProgramException();
+                                    }
+                                    break;
+                            }
+
+                            LayoutInt[] layout = (isGcField ? gcStatics : nonGcStatics);
+                            LayoutInt offset = LayoutInt.AlignUp(layout[index], new LayoutInt(alignment));
+                            layout[index] = offset + new LayoutInt(size);
+                            FieldDesc fieldDesc = null;
+                            try
+                            {
+                                fieldDesc = module.GetField(fieldDefHandle);
+                                if (fieldsForType == null)
+                                {
+                                    fieldsForType = new List<FieldAndOffset>();
+                                }
+                                fieldsForType.Add(new FieldAndOffset(fieldDesc, offset));
+                            }
+                            catch (Exception)
+                            {
+                                // ignore unresolvable fields
+                            }
+                            if (fieldDesc != null)
+                            {
+                                TypeDesc fieldType = fieldDesc.FieldType;
```

Comment:
Accessing the field type can throw. It might be better to handle it in the big switch block above. IsByRef is `ELEMENT_TYPE_BYREF` so that will already throw, albeit a different exception. And `IsByRefLike` is a custom attribute walk, same as the thread static case.

Reply:
Hopefully fixed.

Reply:
Hmm, now I see the IsByRefLike check is actually incorrect (it looks at the declaring type, not at the field type). I'll fix this after the meeting.

Reply:
Hopefully fixed in the last iteration.